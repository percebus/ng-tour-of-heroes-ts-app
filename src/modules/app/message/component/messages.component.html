<div class="messages-component" *ngIf="messageService.messages.length">
  <h2>Messages</h2>
  <button type="button" class="clear" (click)="clear()">Clear messages</button>
  <div *ngFor="let message of messageService.messages">
    {{ message }}
  </div>
</div>
