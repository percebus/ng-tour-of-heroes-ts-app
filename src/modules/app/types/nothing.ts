export type Nullable<T> = T | null;

export type Optional<T> = T | undefined;

export type Tentative<T> = T | undefined | null;
