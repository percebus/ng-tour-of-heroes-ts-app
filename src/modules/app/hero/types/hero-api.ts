import { Observable } from 'rxjs';
import { Optional } from '../../types/nothing';
import { Hero } from './hero';

export interface HeroAPI {
  post(hero: Hero): Observable<Hero>;
  put(hero: Hero): Observable<any>; // FIXME
  delete(hero: Hero): Observable<Hero>;
  getById(id: number): Observable<Optional<Hero>>;
  getAll(): Observable<Hero[]>;
}
