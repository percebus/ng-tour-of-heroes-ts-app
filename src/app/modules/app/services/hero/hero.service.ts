import { Observable, of } from 'rxjs';
import { Injectable } from '@angular/core';
import { MessageService } from '../message/message.service';
import { Hero } from 'src/app/types/hero';
import { heroes } from 'src/data/mock/heroes';

@Injectable({
  providedIn: 'root',
})
export class HeroService {
  constructor(private messageService: MessageService) {}

  getAll(): Observable<Array<Hero>> {
    this.messageService.add('HeroService: fetched heroes');
    return of(heroes);
  }
}
