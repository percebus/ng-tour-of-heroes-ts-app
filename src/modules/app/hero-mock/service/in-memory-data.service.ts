import { Injectable } from '@angular/core';
import { InMemoryDbService } from 'angular-in-memory-web-api';
import { Hero } from '../../hero/types/hero';
import { heroes } from '../data/heroes';

@Injectable({
  providedIn: 'root',
})
export class InMemoryDataService implements InMemoryDbService {
  createDb() {
    return { heroes };
  }

  // Overrides the genId method to ensure that a hero always has an id.
  // If the heroes array is empty,
  // the method below returns the initial number (11).
  // if the heroes array is not empty, the method below returns the highest
  // hero id + 1.
  genId(heroes: Hero[]): number {
    const id =
      heroes.length > 0 ? Math.max(...heroes.map((oHero) => oHero.id)) : 10;

    return id + 1;
  }
}
