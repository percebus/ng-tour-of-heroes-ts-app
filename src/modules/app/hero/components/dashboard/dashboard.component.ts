import { Component, OnInit } from '@angular/core';
import { Hero } from '../../types/hero';
import { HeroService } from '../../service/hero.service';

@Component({
  selector: 'app-dashboard',
  templateUrl: './dashboard.component.html',
  styleUrls: ['./dashboard.component.scss'],
})
export class DashboardComponent implements OnInit {
  heroes: Hero[] = [];

  constructor(private heroService: HeroService) {}

  refresh(): void {
    this.heroService
      .getAll() // Observable< Hero[] >
      .subscribe((heroes) => (this.heroes = heroes.slice(1, 5)));
  }

  ngOnInit(): void {
    this.refresh();
  }
}
