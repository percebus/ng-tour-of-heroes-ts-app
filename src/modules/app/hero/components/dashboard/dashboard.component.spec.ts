import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { HeroAPI } from '../../types/hero-api';
import { MockHeroService } from '../../../hero-mock/services/mock/hero.mock-service';
import { HeroSearchComponent } from '../search/hero-search.component';
import { DashboardComponent } from './dashboard.component';
import { HeroService } from '../../service/hero.service';

describe('DashboardComponent', () => {
  let oDashboardComponent: DashboardComponent;
  let oComponentFixture: ComponentFixture<DashboardComponent>;
  let mockHeroService: HeroAPI;
  let refresh: jest.SpyInstance;

  describe('using MockHeroService', () => {
    beforeEach(() => {
      TestBed.configureTestingModule({
        imports: [HttpClientTestingModule],
        declarations: [DashboardComponent, HeroSearchComponent],
      })
        .overrideComponent(DashboardComponent, {
          set: {
            providers: [{ provide: HeroService, useClass: MockHeroService }],
          },
        })
        .compileComponents();

      oComponentFixture = TestBed.createComponent(DashboardComponent);
      mockHeroService =
        oComponentFixture.debugElement.injector.get(HeroService);

      oDashboardComponent = oComponentFixture.componentInstance;
      refresh = jest.spyOn(oDashboardComponent, 'refresh');

      oComponentFixture.detectChanges();
    });

    it('is instanceof DashboardComponent', () => {
      expect(oDashboardComponent).toBeInstanceOf(DashboardComponent);
    });

    describe('Dependency Injection', () => {
      describe('HeroService', () => {
        it('is instanceof MockHeroService', () => {
          expect(mockHeroService).not.toBeInstanceOf(HeroService);
          expect(mockHeroService).toBeInstanceOf(MockHeroService);
        });
      });
    });

    describe('.refresh()', () => {
      it('got invoked OnInit', () => {
        expect(refresh).toHaveBeenCalled();
      });
    });

    describe('HTML', () => {
      describe('heading', () => {
        it('renders "Top Heroes"', () => {
          const oHTMLElement = oComponentFixture.nativeElement //
            .querySelector('.heroes-dashboard');

          expect(oHTMLElement.textContent).toContain('Top Heroes');
        });
      });

      describe('links', () => {
        let oHTMLElement: HTMLElement;
        beforeEach(() => {
          oHTMLElement =
            oComponentFixture.nativeElement.querySelector('.heroes-menu');
        });

        it('contains 4 heroes', () => {
          expect(oHTMLElement.children.length).toEqual(4);
        });

        it('contains "Bombasto', () => {
          expect(oHTMLElement.textContent).toContain('Bombasto');
        });

        it('contains "Celeritas', () => {
          expect(oHTMLElement.textContent).toContain('Celeritas');
        });

        it('contains "Magneta', () => {
          expect(oHTMLElement.textContent).toContain('Magneta');
        });

        it('contains "RubberMan', () => {
          expect(oHTMLElement.textContent).toContain('RubberMan');
        });
      });

      describe('HeroSearchComponent', () => {
        it('renders HTML into <app-search-hero />"', () => {
          const oHTMLElement = oComponentFixture.nativeElement //
            .querySelector('#search-component');

          expect(oHTMLElement.textContent).toContain('Hero Search');
        });
      });

      it('matches snapshot', () => {
        expect(oComponentFixture).toMatchSnapshot();
      });
    });
  });
});
