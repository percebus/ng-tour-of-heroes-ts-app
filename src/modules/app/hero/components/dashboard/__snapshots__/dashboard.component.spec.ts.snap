// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardComponent using MockHeroService rendering matches snapshot 1`] = `
<app-dashboard
  heroService={[Function MockHeroService]}
  heroes={[Function Array]}
  refresh={[Function Function]}
>
  <div
    class="heroes-dashboard"
  >
    <h2>
      Top Heroes
    </h2>
    <div
      class="heroes-menu"
    >
      <a>
         Bombasto 
      </a>
      <a>
         Celeritas 
      </a>
      <a>
         Magneta 
      </a>
      <a>
         RubberMan 
      </a>
    </div>
    <app-hero-search>
      <div
        id="search-component"
      >
        <label
          for="search-box"
        >
          Hero Search
        </label>
        <input
          id="search-box"
        />
        <ul
          class="search-result"
        >
        </ul>
      </div>
    </app-hero-search>
  </div>
</app-dashboard>
`;
